File: python/string/3120-Count-the-number-of-special-characters-i.py
# Problem: Leetcode 3120 - Count the Number of Special Characters I
# Difficulty: Easy
# Link: https://leetcode.com/problems/count-the-number-of-special-characters-i/description/
# Time Complexity: O(n) as we loop through the set made from string
# Space Complexity: O(n) as we use a set which can be of length n
# Approach: Simple approach that since we have to check 'membership' we loop through the set and we check if a character is lowercase then does 
# its uppercase version is also in the set. Checking in string or list would make this O(n2) but in a set its O(n) overall.

class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        word = set(word)
        cnt = 0

        for char in word:
            if char.islower() and char.upper() in word:
                cnt+=1

        return cnt
        

File: Java/3120-Count-the-number-of-special-characters-i.java
/* 
Problem: Leetcode 3120 - Count the number of special characters in a string
# Difficulty: Medium
# Link: https://leetcode.com/problems/count-the-number-of-special-characters-in-a-string/description/
# Time Complexity: O(n) - as we go through all the elements of the string in worst case
# Space Complexity: O(k) as we use a HashSet to store unique characters, where k is the number of unique characters
# Approach: We iterate through the string and add each character to a HashSet. Then we iterate through the set and check if a lowercase character has its uppercase counterpart in the set. 
# If so, we increment the count.
*/
package Java;
import java.util.Set;
import java.util.HashSet;

class Solution {
    public int numberOfSpecialChars(String word) {
        Set<Character> wordSet = new HashSet<>();
        for (char c:word.toCharArray()){wordSet.add(c);}
        int cnt = 0;
        for (char s:wordSet){
            if (Character.isLowerCase(s) && 
            wordSet.contains(Character.toUpperCase(s))) {
                cnt++;
                }

        }
        return cnt;
    }
}